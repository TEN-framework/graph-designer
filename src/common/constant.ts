const BASE_URL = process.env.NEXT_PUBLIC_BASE_URL || "http://localhost"
const PORT = process.env.NEXT_PUBLIC_PORT || 49483
const PREFIX = "/api/dev-server/v1"


export const DEFAULT_APP = "localhost"
export const DEFAULT_EXTENTION_GROUP = "default"
export const API_URL = `${BASE_URL}:${PORT}${PREFIX}`
