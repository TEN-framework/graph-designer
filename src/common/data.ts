export class EditorData {
  // extensionGroup =>  {
  //  [extensionName]: nodeId
  // }
  nodeMap: Map<string, { [extensionName: string]: string }> = new Map()
  nodeId: number = 1
  edgeId: number = 1

  constructor() { }

  genEdgeId() {
    return `${this.edgeId++}`
  }

  genNodeId() {
    return `${this.nodeId++}`
  }

  getNodeId(extensionGroup: string, extensionname: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (data) {
      if (!data[extensionname]) {
        throw new Error(
          `Invalid extension Node: ${extensionname}, not found in extensionGroup: ${extensionGroup}`,
        )
      }
      return data[extensionname]
    }
    throw new Error(
      `Invalid extensionGroup: ${extensionGroup}, not found in nodeMap`,
    )
  }

  delNode(extensionGroup: string, extensionname: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (data) {
      delete data[extensionname]
    }
  }

  saveNodeId(extensionGroup: string, extensionname: string, nodeId: string) {
    let data = this.nodeMap.get(extensionGroup)
    if (!data) {
      data = {}
      this.nodeMap.set(extensionGroup, data)
    }
    data[extensionname] = nodeId
  }

  clear() {
    this.nodeMap.clear()
    this.nodeId = 1
    this.edgeId = 1
  }
}


export const editorData = new EditorData()

if (process.env.NODE_ENV === "development") {
  if (typeof window !== "undefined") {
    // @ts-ignore
    window.editorData = editorData
  }
}
