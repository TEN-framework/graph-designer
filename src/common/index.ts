export * from "./logger"
export * from "./utils"
export * from "./request"
export * from "./hooks"
export * from "./mock"
export * from "./event"
export * from "./constant"
