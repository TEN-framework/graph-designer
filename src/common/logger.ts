import { formatTime } from "./utils"
import { LogLevel } from "../types"

export interface LoggerConfig {
  level?: LogLevel
  prefix?: string
}

class Logger {
  level: LogLevel = LogLevel.ERROR
  prefix?: string = ""
  preTime?: number = 0

  constructor(config: LoggerConfig) {
    const { level, prefix } = config
    if (level !== undefined) {
      this.level = level
    }
    if (prefix) {
      this.prefix = prefix
    }
  }

  setLogLevel(level: LogLevel) {
    this.level = level
  }

  debug(...args: any[]) {
    this.level <= LogLevel.DEBUG &&
      this._log(`${this._genPrefix()}[DEBUG]: `, ...args)
  }

  warn(...args: any[]) {
    this.level <= LogLevel.WARN &&
      this._warn(`${this._genPrefix()}[WARN]: `, ...args)
  }

  error(...args: any[]) {
    this.level <= LogLevel.ERROR &&
      this._err(`${this._genPrefix()}[ERROR]: `, ...args)
  }

  time(...args: any[]) {
    const time = new Date().getTime()
    let cost = 0
    let start = ""
    if (this.preTime) {
      cost = time - this.preTime
    } else {
      start = "start"
    }
    this.preTime = time
    this._log(
      `${this._genPrefix()}[time]:  -------------- cost:${cost}ms ${start} -----------------   \n`,
      ...args,
    )
  }

  timeEnd(...args: any[]) {
    const time = new Date().getTime()
    let cost = 0
    if (this.preTime) {
      cost = time - this.preTime
    }
    this.preTime = 0
    this._log(
      `${this._genPrefix()}[time]:  -------------- cost:${cost}ms end -----------------   \n`,
      ...args,
    )
  }

  //  ---------------------------- private ----------------------------

  private _log(...args: any[]) {
    const [arg, ...other] = args
    console.log(`%c${arg}`, "color:yellow", ...other)
  }

  private _warn(...args: any[]) {
    const [arg, ...other] = args
    console.warn(`%c${arg}`, "color:yellow", ...other)
  }

  private _err(...args: any[]) {
    let res = ""
    for (const item of args) {
      res += typeof item == "string" ? item : JSON.stringify(item)
    }
    console.error(res)
  }

  private _genPrefix() {
    return formatTime() + (this.prefix ? ` [${this.prefix}]` : "")
  }
}

export const logger = new Logger({
  level: LogLevel.DEBUG,
  prefix: "editor",
})

export default Logger
