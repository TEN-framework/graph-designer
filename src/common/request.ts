import { message } from "antd"
import { camelToSnake } from "./utils"
import {
  IConnection,
  IExtension,
  ICompatibleConnection,
  IGraphData,
} from "@/types"
import { DEFAULT_EXTENTION_GROUP, API_URL } from "./constant"

export const apiGetVersion = async () => {
  const res = await fetch(`${API_URL}/version`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get version")
  }
  return data?.data ?? {}
}

export const apiGetInstalledExtension = async () => {
  const res = await fetch(`${API_URL}/addons/extensions`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get installed extensions")
  }
  let arr = data?.data ?? []

  return arr.map((item: IExtension) => {
    return {
      ...item,
      addon: item.name,
      app: "localhost",
      extension_group: DEFAULT_EXTENTION_GROUP,
      property: null,
    }
  })
}

export const apiAllGetGraph = async () => {
  const res = await fetch(`${API_URL}/graphs`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get all graphs")
  }
  return data?.data ?? []
}

export const apiGetGraphExtension = async (
  graphName: string,
): Promise<IExtension[]> => {
  const res = await fetch(`${API_URL}/graphs/${graphName}/extensions`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get graph extensions")
  }
  return (data?.data ?? []) as IExtension[]
}

export const apiGetGraphConnection = async (
  graphName: string,
): Promise<IConnection[]> => {
  const res = await fetch(`${API_URL}/graphs/${graphName}/connections`, {
    method: "GET",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to get graph connections")
  }
  return (data?.data ?? []) as IConnection[]
}

export const apiQueryCompatibleMessage = async (
  options: ICompatibleConnection,
): Promise<ICompatibleConnection[]> => {
  const res = await fetch(`${API_URL}/messages/compatible`, {
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
    method: "POST",
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error(data.message)
  }
  let arr = data?.data ?? []
  return arr as ICompatibleConnection[]
}

export const apiUpdateGraph = async (
  graphName: string,
  graphData: IGraphData,
) => {
  const res = await fetch(`${API_URL}/graphs/${graphName}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(graphData),
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to update graph")
  }
  return data
}

export const apiSaveGraph = async () => {
  const res = await fetch(`${API_URL}/manifest`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
  })
  const data = await res.json()
  if (data.status != "ok") {
    throw new Error("Failed to save graph")
  }
  return data
}
