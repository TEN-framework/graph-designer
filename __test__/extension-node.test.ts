import { expect, test, describe } from 'vitest'
import { MOCK_EXTENTIONS, extensionsToNodes, nodesToExtensions, MOCK_INSTALLED_EXTENTIONS } from "@/common"
import type { ExtensionNode } from "@/types"

let nodes: ExtensionNode[]

describe("extension and node convert each other", () => {

  test('extensions to nods', () => {
    nodes = extensionsToNodes(MOCK_EXTENTIONS)
    expect(nodes).toMatchSnapshot();
  })

  test('nodes to extensions', () => {
    const extensions = nodesToExtensions(nodes, MOCK_INSTALLED_EXTENTIONS)
    expect(extensions).toMatchSnapshot();
  })


})



