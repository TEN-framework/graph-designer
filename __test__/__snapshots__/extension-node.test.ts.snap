// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extension and node convert each other > extensions to nods 1`] = `
[
  {
    "data": {
      "addon": "agora_rtc",
      "extensionGroup": "default",
      "inputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "data",
          "status": "default",
          "type": "data",
        },
        {
          "name": "pcm_frame",
          "status": "default",
          "type": "pcm_frame",
        },
        {
          "name": "image_frame",
          "status": "default",
          "type": "image_frame",
        },
      ],
      "name": "agora_rtc",
      "outputs": [
        {
          "name": "on_user_joined",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "on_user_left",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "on_connection_failure",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
        {
          "name": "pcm_frame",
          "status": "default",
          "type": "pcm_frame",
        },
        {
          "name": "image_frame",
          "status": "default",
          "type": "image_frame",
        },
      ],
      "property": {
        "agora_asr_language": "en-US",
        "agora_asr_session_control_file_path": "session_control.conf",
        "agora_asr_vendor_key": "<azure_stt_key>",
        "agora_asr_vendor_name": "microsoft",
        "agora_asr_vendor_region": "<azure_stt_region>",
        "app_id": "<agora_appid>",
        "channel": "astra_agents_test",
        "enable_agora_asr": true,
        "publish_audio": true,
        "publish_data": true,
        "remote_stream_id": 123,
        "stream_id": 1234,
        "subscribe_audio": true,
        "token": "<agora_token>",
      },
      "propertyTypes": {
        "agora_asr_language": {
          "type": "string",
        },
        "agora_asr_session_control_file_path": {
          "type": "string",
        },
        "agora_asr_vendor_key": {
          "type": "string",
        },
        "agora_asr_vendor_name": {
          "type": "string",
        },
        "agora_asr_vendor_region": {
          "type": "string",
        },
        "app_id": {
          "type": "string",
        },
        "area_code": {
          "type": "int32",
        },
        "channel": {
          "type": "string",
        },
        "enable_agora_asr": {
          "type": "bool",
        },
        "encryption_key": {
          "type": "string",
        },
        "encryption_mode": {
          "type": "int32",
        },
        "encryption_salt": {
          "type": "string",
        },
        "publish_audio": {
          "type": "bool",
        },
        "publish_data": {
          "type": "bool",
        },
        "publish_video": {
          "type": "bool",
        },
        "publish_video_frame_rate": {
          "type": "int32",
        },
        "publish_video_height": {
          "type": "int32",
        },
        "publish_video_width": {
          "type": "int32",
        },
        "remote_stream_id": {
          "type": "Uint32",
        },
        "sdk_log_file_path": {
          "type": "string",
        },
        "stream_id": {
          "type": "Uint32",
        },
        "subscribe_audio": {
          "type": "bool",
        },
        "subscribe_audio_num_of_channels": {
          "type": "int32",
        },
        "subscribe_audio_sample_rate": {
          "type": "int32",
        },
        "subscribe_video": {
          "type": "bool",
        },
        "subscribe_video_pix_fmt": {
          "type": "int32",
        },
        "token": {
          "type": "string",
        },
      },
      "status": "default",
    },
    "id": "1",
    "position": {
      "x": 0,
      "y": 0,
    },
    "sourcePosition": "right",
    "targetPosition": "left",
    "type": "extension",
  },
  {
    "data": {
      "addon": "interrupt_detector",
      "extensionGroup": "default",
      "inputs": [
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
      ],
      "name": "interrupt_detector",
      "outputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
      ],
      "property": null,
      "propertyTypes": undefined,
      "status": "default",
    },
    "id": "2",
    "position": {
      "x": 0,
      "y": 0,
    },
    "sourcePosition": "right",
    "targetPosition": "left",
    "type": "extension",
  },
  {
    "data": {
      "addon": "openai_chatgpt",
      "extensionGroup": "chatgpt",
      "inputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
      ],
      "name": "openai_chatgpt",
      "outputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
      ],
      "property": {
        "api_key": "<openai_api_key>",
        "base_url": "",
        "frequency_penalty": 0.9,
        "greeting": "ASTRA agent connected. How can i help you today?",
        "max_memory_length": 10,
        "max_tokens": 512,
        "model": "gpt-3.5-turbo",
        "prompt": "",
        "proxy_url": "",
      },
      "propertyTypes": {
        "api_key": {
          "type": "string",
        },
        "frequency_penalty": {
          "type": "float64",
        },
        "greeting": {
          "type": "string",
        },
        "max_memory_length": {
          "type": "int64",
        },
        "max_tokens": {
          "type": "int64",
        },
        "model": {
          "type": "string",
        },
        "presence_penalty": {
          "type": "float64",
        },
        "prompt": {
          "type": "string",
        },
      },
      "status": "default",
    },
    "id": "3",
    "position": {
      "x": 0,
      "y": 0,
    },
    "sourcePosition": "right",
    "targetPosition": "left",
    "type": "extension",
  },
  {
    "data": {
      "addon": "azure_tts",
      "extensionGroup": "tts",
      "inputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
      ],
      "name": "azure_tts",
      "outputs": [
        {
          "name": "flush",
          "status": "default",
          "type": "cmd",
        },
        {
          "name": "pcm_frame",
          "status": "default",
          "type": "pcm_frame",
        },
      ],
      "property": {
        "azure_subscription_key": "<azure_tts_key>",
        "azure_subscription_region": "<azure_tts_region>",
        "azure_synthesis_voice_name": "en-US-JaneNeural",
      },
      "propertyTypes": {
        "azure_subscription_key": {
          "type": "string",
        },
        "azure_subscription_region": {
          "type": "string",
        },
        "azure_synthesis_voice_name": {
          "type": "string",
        },
      },
      "status": "default",
    },
    "id": "4",
    "position": {
      "x": 0,
      "y": 0,
    },
    "sourcePosition": "right",
    "targetPosition": "left",
    "type": "extension",
  },
  {
    "data": {
      "addon": "chat_transcriber",
      "extensionGroup": "transcriber",
      "inputs": [
        {
          "name": "text_data",
          "status": "default",
          "type": "data",
        },
      ],
      "name": "chat_transcriber",
      "outputs": [
        {
          "name": "data",
          "status": "default",
          "type": "data",
        },
      ],
      "property": null,
      "propertyTypes": undefined,
      "status": "default",
    },
    "id": "5",
    "position": {
      "x": 0,
      "y": 0,
    },
    "sourcePosition": "right",
    "targetPosition": "left",
    "type": "extension",
  },
]
`;

exports[`extension and node convert each other > nodes to extensions 1`] = `
[
  {
    "addon": "agora_rtc",
    "api": {
      "cmd_in": [
        {
          "name": "flush",
        },
      ],
      "cmd_out": [
        {
          "name": "on_user_joined",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "channel",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "user_id",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "remote_user_id",
            },
          ],
        },
        {
          "name": "on_user_left",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "channel",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "user_id",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "remote_user_id",
            },
          ],
        },
        {
          "name": "on_connection_failure",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "channel",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "user_id",
            },
          ],
        },
      ],
      "data_in": [
        {
          "name": "data",
        },
      ],
      "data_out": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "int64",
              },
              "name": "time",
            },
            {
              "attributes": {
                "type": "int64",
              },
              "name": "duration_ms",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "language",
            },
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "is_final",
            },
            {
              "attributes": {
                "type": "Uint32",
              },
              "name": "stream_id",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "end_of_segment",
            },
          ],
        },
      ],
      "video_frame_in": [
        {
          "name": "image_frame",
        },
      ],
      "video_frame_out": [
        {
          "name": "image_frame",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "channel",
            },
            {
              "attributes": {
                "type": "Uint32",
              },
              "name": "stream_id",
            },
          ],
        },
      ],
      "audio_frame_in": [
        {
          "name": "pcm_frame",
        },
      ],
      "audio_frame_out": [
        {
          "name": "pcm_frame",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "channel",
            },
            {
              "attributes": {
                "type": "Uint32",
              },
              "name": "stream_id",
            },
          ],
        },
      ],
      "property": {
        "agora_asr_language": {
          "type": "string",
        },
        "agora_asr_session_control_file_path": {
          "type": "string",
        },
        "agora_asr_vendor_key": {
          "type": "string",
        },
        "agora_asr_vendor_name": {
          "type": "string",
        },
        "agora_asr_vendor_region": {
          "type": "string",
        },
        "app_id": {
          "type": "string",
        },
        "area_code": {
          "type": "int32",
        },
        "channel": {
          "type": "string",
        },
        "enable_agora_asr": {
          "type": "bool",
        },
        "encryption_key": {
          "type": "string",
        },
        "encryption_mode": {
          "type": "int32",
        },
        "encryption_salt": {
          "type": "string",
        },
        "publish_audio": {
          "type": "bool",
        },
        "publish_data": {
          "type": "bool",
        },
        "publish_video": {
          "type": "bool",
        },
        "publish_video_frame_rate": {
          "type": "int32",
        },
        "publish_video_height": {
          "type": "int32",
        },
        "publish_video_width": {
          "type": "int32",
        },
        "remote_stream_id": {
          "type": "Uint32",
        },
        "sdk_log_file_path": {
          "type": "string",
        },
        "stream_id": {
          "type": "Uint32",
        },
        "subscribe_audio": {
          "type": "bool",
        },
        "subscribe_audio_num_of_channels": {
          "type": "int32",
        },
        "subscribe_audio_sample_rate": {
          "type": "int32",
        },
        "subscribe_video": {
          "type": "bool",
        },
        "subscribe_video_pix_fmt": {
          "type": "int32",
        },
        "token": {
          "type": "string",
        },
      },
    },
    "app": "localhost",
    "extension_group": "default",
    "name": "agora_rtc",
    "property": {
      "agora_asr_language": "en-US",
      "agora_asr_session_control_file_path": "session_control.conf",
      "agora_asr_vendor_key": "<azure_stt_key>",
      "agora_asr_vendor_name": "microsoft",
      "agora_asr_vendor_region": "<azure_stt_region>",
      "app_id": "<agora_appid>",
      "channel": "astra_agents_test",
      "enable_agora_asr": true,
      "publish_audio": true,
      "publish_data": true,
      "remote_stream_id": 123,
      "stream_id": 1234,
      "subscribe_audio": true,
      "token": "<agora_token>",
    },
  },
  {
    "addon": "interrupt_detector",
    "api": {
      "cmd_out": [
        {
          "name": "flush",
        },
      ],
      "data_in": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "is_final",
            },
          ],
        },
      ],
    },
    "app": "localhost",
    "extension_group": "default",
    "name": "interrupt_detector",
    "property": null,
  },
  {
    "addon": "openai_chatgpt",
    "api": {
      "cmd_in": [
        {
          "name": "flush",
        },
      ],
      "cmd_out": [
        {
          "name": "flush",
        },
      ],
      "data_in": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "is_final",
            },
          ],
        },
      ],
      "data_out": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "end_of_segment",
            },
          ],
        },
      ],
      "property": {
        "api_key": {
          "type": "string",
        },
        "frequency_penalty": {
          "type": "float64",
        },
        "greeting": {
          "type": "string",
        },
        "max_memory_length": {
          "type": "int64",
        },
        "max_tokens": {
          "type": "int64",
        },
        "model": {
          "type": "string",
        },
        "presence_penalty": {
          "type": "float64",
        },
        "prompt": {
          "type": "string",
        },
      },
    },
    "app": "localhost",
    "extension_group": "chatgpt",
    "name": "openai_chatgpt",
    "property": {
      "api_key": "<openai_api_key>",
      "base_url": "",
      "frequency_penalty": 0.9,
      "greeting": "ASTRA agent connected. How can i help you today?",
      "max_memory_length": 10,
      "max_tokens": 512,
      "model": "gpt-3.5-turbo",
      "prompt": "",
      "proxy_url": "",
    },
  },
  {
    "addon": "azure_tts",
    "api": {
      "cmd_in": [
        {
          "name": "flush",
        },
      ],
      "cmd_out": [
        {
          "name": "flush",
        },
      ],
      "data_in": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
          ],
        },
      ],
      "audio_frame_out": [
        {
          "name": "pcm_frame",
        },
      ],
      "property": {
        "azure_subscription_key": {
          "type": "string",
        },
        "azure_subscription_region": {
          "type": "string",
        },
        "azure_synthesis_voice_name": {
          "type": "string",
        },
      },
    },
    "app": "localhost",
    "extension_group": "tts",
    "name": "azure_tts",
    "property": {
      "azure_subscription_key": "<azure_tts_key>",
      "azure_subscription_region": "<azure_tts_region>",
      "azure_synthesis_voice_name": "en-US-JaneNeural",
    },
  },
  {
    "addon": "chat_transcriber",
    "api": {
      "data_in": [
        {
          "name": "text_data",
          "property": [
            {
              "attributes": {
                "type": "string",
              },
              "name": "text",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "is_final",
            },
            {
              "attributes": {
                "type": "Uint32",
              },
              "name": "stream_id",
            },
            {
              "attributes": {
                "type": "bool",
              },
              "name": "end_of_segment",
            },
          ],
        },
      ],
      "data_out": [
        {
          "name": "data",
        },
      ],
    },
    "app": "localhost",
    "extension_group": "transcriber",
    "name": "chat_transcriber",
    "property": null,
  },
]
`;
